### PR TITLE
Update @typescript-eslint package versions and update TypeScript rules

### PR DESCRIPTION
**Pull request by jsonunger**
The Life Growth team is looking to move to TypeScript 4.0, but are currently blocked by issues with the ESLint configuration for TypeScript not yet supporting that version. I checked the `@typescript-eslint` repository and the latest version supports from 3.3 up to 4.1. From my check, we do not have any repositories using a TypeScript version older than 3.5, so this would still be supported.

Additionally, I saw that the `@typescript-eslint/eslint-plugin` has a variety of rules that ["extend"](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#extension-rules) the main rules while adding TypeScript processing, so I went through the base configuration `/index.js` and all of its extended configs and added as many of the extended rules as possible until things broke for our repository. All of the rules that I added are compatible with version 5, so I didn't need to change the peer or dev dependency.

### CODE REVIEW

File: typescript.js
Comment (jmcolella) [inline typescript.js:13]: I just want to double-check: it was required to override the original ESLint rules for the TS rules to take effect?

In addition, was there a reason for putting the TS rules in `overrides` as opposed to the top-level `rules` config?

Comment (jsonunger) [reply typescript.js:13]: 1. Just using [no-unused-vars](https://github.com/typescript-eslint/typescript-eslint/blob/3e204e9f2fe7008c7d7127d865f58bff1d978e0a/packages/eslint-plugin/docs/rules/no-unused-vars.md) as an example, the extension rules all tell you to disable the base rule.

2. Hypothetically there is no reason. We were previously using overrides for some but not all, so I put them all as overrides for consistency. However, I am happy to defer to the judgement of the stewards if you think they should be top-level instead, since there shouldn't be any reason they couldn't all be top-level as all of the extension rules maintain the functionality of the base rules.

Comment (jmcolella) [reply typescript.js:13]: 1. 🤦 I should have seen that, thanks for pointing out!
2. Gotcha, thanks for the context! I think this is fine, a good benefit I see is co-locating the ESLint rule with the TS rule so it's a bit easier to understand that we are turning 1 off in favor of the other

**Review by jmcolella — APPROVED** [review]
@jsonunger LGTM! Let me know if you need help publishing to NPM and myself and @mttrms can assist

File: typescript.js
Comment (mttrms) [reply typescript.js:13]: yeah I like the idea of having them together - thanks for adding them this way.

**Review by mttrms — APPROVED** [review]
LGTM!

**Comment by jsonunger** [comment]
@jmcolella @mttrms Moved most of the rules to be top-level after testing locally and seeing no real difference on the `life-web` codebase either in terms of timing or the result of running lint on the codebase.

**Comment by jmcolella** [comment]
@jsonunger sounds good to me

**Pull request merged**
